File: lsrc_web_application/src/service/chapterService.ts
// src/service/chapterService.ts
import api from '../api/axiosConfig';
import type { Chapter, ChapterRequest, UpdateOrderRequest } from '../types/chapter.types';
import type { RequestResponse } from '../types/course.types';

/**
 * Lấy danh sách chương của khóa học (kèm lessons)
 */
export const getChaptersByCourse = async (courseId: number): Promise<Chapter[]> => {
  const response = await api.get<RequestResponse<Chapter[]>>(`/courses/${courseId}/chapters`);
  return response.data.data;
};

/**
 * Lấy chi tiết 1 chương
 */
export const getChapterById = async (chapterId: number): Promise<Chapter> => {
  const response = await api.get<RequestResponse<Chapter>>(`/chapters/${chapterId}`);
  return response.data.data;
};

/**
 * Lấy chương kèm bài học
 */
export const getChapterWithLessons = async (chapterId: number): Promise<Chapter> => {
  const response = await api.get<RequestResponse<Chapter>>(`/chapters/${chapterId}/lessons`);
  return response.data.data;
};

/**
 * Tạo chương mới
 */
export const createChapter = async (courseId: number, data: ChapterRequest): Promise<Chapter> => {
  const response = await api.post<RequestResponse<Chapter>>(`/courses/${courseId}/chapters`, data);
  return response.data.data;
};

/**
 * Cập nhật chương
 */
export const updateChapter = async (chapterId: number, data: ChapterRequest): Promise<Chapter> => {
  const response = await api.put<RequestResponse<Chapter>>(`/chapters/${chapterId}`, data);
  return response.data.data;
};

/**
 * Xóa chương
 */
export const deleteChapter = async (chapterId: number): Promise<void> => {
  await api.delete(`/chapters/${chapterId}`);
};

/**
 * Đổi thứ tự 1 chương
 */
export const updateChapterOrder = async (chapterId: number, newOrderIndex: number): Promise<void> => {
  await api.patch(`/chapters/${chapterId}/order`, { newOrderIndex });
};

/**
 * Cập nhật thứ tự hàng loạt
 */
export const updateChapterOrderBatch = async (
  updates: UpdateOrderRequest[]
): Promise<void> => {
  await api.put('/chapters/batch/order', updates);
};

File: OC01/src/main/java/org/wisdom/oc01/service/ChapterService.java
// ============================================
// ChapterService.java - Service Interface
// ============================================
package org.wisdom.oc01.service;

import org.wisdom.oc01.dto.request.ChapterRequest;
import org.wisdom.oc01.dto.request.UpdateOrderRequest;
import org.wisdom.oc01.dto.response.ChapterResponse;
import java.util.List;

public interface ChapterService {
    // ==================== GET ====================

    /** Lấy chương của khóa học */
    List<ChapterResponse> getChaptersByCourse(Integer courseId);

    /** Lấy chương theo ID */
    ChapterResponse getChapterById(Integer chapterId);

    /** Lấy chương kèm bài học */
    ChapterResponse getChapterWithLessons(Integer chapterId);

    // ==================== CREATE ====================

    /** Tạo chương mới */
    ChapterResponse createChapter(Integer courseId, ChapterRequest request);

    // ==================== UPDATE ====================

    /** Cập nhật chương */
    ChapterResponse updateChapter(Integer chapterId, ChapterRequest request);

    /** Cập nhật thứ tự chương */
    ChapterResponse updateOrderIndex(Integer chapterId, Integer newOrderIndex);

    /** Cập nhật thứ tự hàng loạt */
    void updateOrderBatch(List<UpdateOrderRequest> updates);

    // ==================== DELETE ====================

    /** Xóa chương */
    void deleteChapter(Integer chapterId);
}